### PR TITLE
[WIP] prevent negative values in dmu_zfetch

### PR DESCRIPTION
**Pull request by ofaaland**
### Motivation and Context

On our systems, running Lustre on ZFS, we have seen dmu_zfetch() set max_dist to negative values, resulting in dmu_zfetch() not exiting in a reasonable time, and the system being powered off by our HA system.

This patch is to apply to our local build to gather more debugging information, and to prevent those values from going negative.  

After the debugging information reveals what causes the negative values, we will write a patch for master.

### Description

Prevent signed values from becoming negative in dmu_zfetch.  Emit debugging information to the console when the condition occurs.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/dmu_zfetch.c [hunk 1]
```diff
@@ -312,7 +329,29 @@ dmu_zfetch(zfetch_t *zf, uint64_t blkid, uint64_t nblks, boolean_t fetch_data)
 		 */
 		pf_ahead_blks = zs->zs_pf_blkid - blkid + nblks;
 		max_blks = max_dist_blks - (pf_start - end_of_access_blkid);
+
+		if (dmu_zfetch_out_of_range == 0 && max_blks < 0) {
+			toss4454_warn("data:max_blks < 0", blkid, nblks,
+			    orig_zs_blkid, orig_zs_pf_blkid, orig_zs_ipf_blkid,
+			    max_dist_blks);
+			dmu_zfetch_out_of_range = 1;
```
Comment (tonyhutter) [inline module/zfs/dmu_zfetch.c]: `dmu_zfetch_out_of_range = 1;` gets set within `toss4454_warn()`.  Did you want to set it in both places?

File: module/zfs/dmu_zfetch.c [hunk 2]
```diff
@@ -196,6 +196,19 @@ dmu_zfetch_stream_create(zfetch_t *zf, uint64_t blkid)
 	list_insert_head(&zf->zf_stream, zs);
 }
 
+int dmu_zfetch_out_of_range = 0;
+
```
Comment (tonyhutter) [inline module/zfs/dmu_zfetch.c]: I'd recommend renaming this function and adding a comment, like:
```
/* Temporary debug print function to run down a case where max_dist is negative */
debug_max_dist_is_negative(char *constraint, uint64_t blkid, uint64_t nblks,
```

File: module/zfs/dmu_zfetch.c (hunk 1 above)
Comment (ofaaland) [reply module/zfs/dmu_zfetch.c]: Nope, fixed.

File: module/zfs/dmu_zfetch.c (hunk 2 above)
Comment (ofaaland) [reply module/zfs/dmu_zfetch.c]: Good idea.  Done.

**Comment by ofaaland** [comment]
Thank you both for your feedback.

**Pull request closed without merging**